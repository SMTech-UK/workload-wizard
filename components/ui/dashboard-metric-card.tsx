import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 