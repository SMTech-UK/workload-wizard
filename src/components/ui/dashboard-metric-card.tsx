import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 