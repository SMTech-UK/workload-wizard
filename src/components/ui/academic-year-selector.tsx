"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 