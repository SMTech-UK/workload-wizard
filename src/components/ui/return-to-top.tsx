"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 