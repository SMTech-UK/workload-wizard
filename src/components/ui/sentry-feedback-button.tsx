"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 