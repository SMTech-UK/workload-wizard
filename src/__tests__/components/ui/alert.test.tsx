import { render, screen } from '@testing-library/react';
import { Alert, AlertTitle, AlertDescription } from '@/components/ui/alert';

describe('Alert Component - Comprehensive UI Testing', () => {
  describe('Basic Rendering and Functionality', () => {
    it('should render alert with basic content', () => {
      render(
        <Alert>
          <AlertTitle>Alert Title</AlertTitle>
          <AlertDescription>This is an alert description.</AlertDescription>
        </Alert>
      );

      expect(screen.getByRole('alert')).toBeInTheDocument();
      expect(screen.getByText('Alert Title')).toBeInTheDocument();
      expect(screen.getByText('This is an alert description.')).toBeInTheDocument();
    });

    it('should render alert without title', () => {
      render(
        <Alert>
          <AlertDescription>This is an alert without a title.</AlertDescription>
        </Alert>
      );

      expect(screen.getByRole('alert')).toBeInTheDocument();
      expect(screen.getByText('This is an alert without a title.')).toBeInTheDocument();
    });

    it('should render alert without description', () => {
      render(
        <Alert>
          <AlertTitle>Alert Title Only</AlertTitle>
        </Alert>
      );

      expect(screen.getByRole('alert')).toBeInTheDocument();
      expect(screen.getByText('Alert Title Only')).toBeInTheDocument();
    });

    it('should render empty alert', () => {
      render(<Alert />);

      expect(screen.getByRole('alert')).toBeInTheDocument();
    });
  });

  describe('Accessibility Testing', () => {
    it('should have proper ARIA role', () => {
      render(
        <Alert>
          <AlertDescription>Test alert</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toBeInTheDocument();
      expect(alert).toHaveAttribute('role', 'alert');
    });

    it('should have proper heading structure', () => {
      render(
        <Alert>
          <AlertTitle>Important Alert</AlertTitle>
          <AlertDescription>This is important information.</AlertDescription>
        </Alert>
      );

      const title = screen.getByText('Important Alert');
      expect(title.tagName).toBe('H5');
      expect(title).toHaveClass('mb-1', 'font-medium', 'leading-none', 'tracking-tight');
    });

    it('should have proper semantic structure', () => {
      render(
        <Alert>
          <AlertTitle>Test Title</AlertTitle>
          <AlertDescription>Test description</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      const title = screen.getByText('Test Title');
      const description = screen.getByText('Test description');

      expect(alert).toContainElement(title);
      expect(alert).toContainElement(description);
      expect(description.tagName).toBe('DIV');
    });
  });

  describe('Variant Testing', () => {
    it('should render default variant correctly', () => {
      render(
        <Alert variant="default">
          <AlertDescription>Default alert</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('bg-background', 'text-foreground');
      expect(alert).toHaveClass('relative', 'w-full', 'rounded-lg', 'border', 'p-4');
    });

    it('should render destructive variant correctly', () => {
      render(
        <Alert variant="destructive">
          <AlertDescription>Destructive alert</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('border-destructive/50', 'text-destructive');
      expect(alert).toHaveClass('relative', 'w-full', 'rounded-lg', 'border', 'p-4');
    });

    it('should use default variant when no variant is specified', () => {
      render(
        <Alert>
          <AlertDescription>No variant specified</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('bg-background', 'text-foreground');
    });
  });

  describe('Responsive Behavior', () => {
    it('should adapt to different screen sizes', () => {
      // Mock mobile screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      });

      render(
        <Alert>
          <AlertTitle>Responsive Alert</AlertTitle>
          <AlertDescription>This alert should be responsive.</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('w-full');
      expect(alert).toHaveClass('p-4');
    });

    it('should maintain proper spacing on different screen sizes', () => {
      // Mock tablet screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 768,
      });

      render(
        <Alert>
          <AlertTitle>Tablet Alert</AlertTitle>
          <AlertDescription>This alert should work on tablets.</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('p-4');
      expect(alert).toHaveClass('rounded-lg');
    });
  });

  describe('Content Structure', () => {
    it('should handle long content gracefully', () => {
      const longContent = 'A'.repeat(500);

      render(
        <Alert>
          <AlertTitle>Long Content Alert</AlertTitle>
          <AlertDescription>{longContent}</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toBeInTheDocument();
      expect(screen.getByText(longContent)).toBeInTheDocument();
    });

    it('should handle special characters in content', () => {
      const specialContent = 'Content with special chars: <>&"\'<>';

      render(
        <Alert>
          <AlertTitle>Special Characters</AlertTitle>
          <AlertDescription>{specialContent}</AlertDescription>
        </Alert>
      );

      expect(screen.getByText(specialContent)).toBeInTheDocument();
    });

    it('should handle HTML content safely', () => {
      const htmlContent = '<strong>Bold text</strong> and <em>italic text</em>';

      render(
        <Alert>
          <AlertDescription>{htmlContent}</AlertDescription>
        </Alert>
      );

      expect(screen.getByText(htmlContent)).toBeInTheDocument();
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle very long titles', () => {
      const longTitle = 'A'.repeat(200);

      render(
        <Alert>
          <AlertTitle>{longTitle}</AlertTitle>
          <AlertDescription>Description</AlertDescription>
        </Alert>
      );

      expect(screen.getByText(longTitle)).toBeInTheDocument();
    });

    it('should handle empty strings', () => {
      render(
        <Alert>
          <AlertTitle></AlertTitle>
          <AlertDescription></AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toBeInTheDocument();
    });

    it('should handle multiple alerts', () => {
      render(
        <div>
          <Alert>
            <AlertDescription>First alert</AlertDescription>
          </Alert>
          <Alert>
            <AlertDescription>Second alert</AlertDescription>
          </Alert>
        </div>
      );

      const alerts = screen.getAllByRole('alert');
      expect(alerts).toHaveLength(2);
      expect(screen.getByText('First alert')).toBeInTheDocument();
      expect(screen.getByText('Second alert')).toBeInTheDocument();
    });
  });

  describe('Snapshot Testing for UI Consistency', () => {
    it('should match snapshot for default alert', () => {
      const { container } = render(
        <Alert>
          <AlertTitle>Test Alert</AlertTitle>
          <AlertDescription>This is a test alert.</AlertDescription>
        </Alert>
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for destructive alert', () => {
      const { container } = render(
        <Alert variant="destructive">
          <AlertTitle>Error Alert</AlertTitle>
          <AlertDescription>This is an error alert.</AlertDescription>
        </Alert>
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for alert without title', () => {
      const { container } = render(
        <Alert>
          <AlertDescription>Alert without title</AlertDescription>
        </Alert>
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Performance Testing', () => {
    it('should handle rapid re-renders efficiently', () => {
      const { rerender } = render(
        <Alert>
          <AlertDescription>Initial alert</AlertDescription>
        </Alert>
      );

      expect(screen.getByText('Initial alert')).toBeInTheDocument();

      // Rerender with different content
      rerender(
        <Alert>
          <AlertDescription>Updated alert</AlertDescription>
        </Alert>
      );

      expect(screen.getByText('Updated alert')).toBeInTheDocument();
      expect(screen.queryByText('Initial alert')).not.toBeInTheDocument();
    });

    it('should handle variant changes efficiently', () => {
      const { rerender } = render(
        <Alert variant="default">
          <AlertDescription>Default variant</AlertDescription>
        </Alert>
      );

      const alert = screen.getByRole('alert');
      expect(alert).toHaveClass('bg-background', 'text-foreground');

      // Change to destructive variant
      rerender(
        <Alert variant="destructive">
          <AlertDescription>Destructive variant</AlertDescription>
        </Alert>
      );

      expect(alert).toHaveClass('border-destructive/50', 'text-destructive');
      expect(alert).not.toHaveClass('bg-background', 'text-foreground');
    });
  });
}); 