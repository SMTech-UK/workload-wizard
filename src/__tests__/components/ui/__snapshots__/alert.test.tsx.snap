// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for alert without title 1`] = `
<div
  class="relative w-full rounded-lg border p-4 [&>svg~*]:pl-7 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 [&>svg]:text-foreground bg-background text-foreground"
  role="alert"
>
  <div
    class="text-sm [&_p]:leading-relaxed"
  >
    Alert without title
  </div>
</div>
`;

exports[`Alert Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for default alert 1`] = `
<div
  class="relative w-full rounded-lg border p-4 [&>svg~*]:pl-7 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 [&>svg]:text-foreground bg-background text-foreground"
  role="alert"
>
  <h5
    class="mb-1 font-medium leading-none tracking-tight"
  >
    Test Alert
  </h5>
  <div
    class="text-sm [&_p]:leading-relaxed"
  >
    This is a test alert.
  </div>
</div>
`;

exports[`Alert Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for destructive alert 1`] = `
<div
  class="relative w-full rounded-lg border p-4 [&>svg~*]:pl-7 [&>svg+div]:translate-y-[-3px] [&>svg]:absolute [&>svg]:left-4 [&>svg]:top-4 border-destructive/50 text-destructive dark:border-destructive [&>svg]:text-destructive"
  role="alert"
>
  <h5
    class="mb-1 font-medium leading-none tracking-tight"
  >
    Error Alert
  </h5>
  <div
    class="text-sm [&_p]:leading-relaxed"
  >
    This is an error alert.
  </div>
</div>
`;
